stores/slabbuffer.go

```diff
@@ -0,0 +1,505 @@
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/object"
+	"gorm.io/gorm"
+	"lukechampine.com/frand"
+)
+
+type SlabBuffer struct {
+	dbID     uint
+	filename string
+	slabKey  object.EncryptionKey
+	maxSize  int64
+
+	dbMu sync.Mutex
+
+	mu          sync.Mutex
+	file        *os.File
+	lockedUntil time.Time
+	size        int64
+	dbSize      int64
+	syncErr     error
+}
+
+type bufferGroupID [6]byte
+
+type SlabBufferManager struct {
+	bufferedSlabCompletionThreshold int64
+	dir                             string
+	s                               *SQLStore
+
+	mu                sync.Mutex
+	completeBuffers   map[bufferGroupID][]*SlabBuffer
+	incompleteBuffers map[bufferGroupID][]*SlabBuffer
+	buffersByKey      map[string]*SlabBuffer
+}
+
+func newSlabBufferManager(sqlStore *SQLStore, slabBufferCompletionThreshold int64, partialSlabDir string) (*SlabBufferManager, error) {
+	if slabBufferCompletionThreshold < 0 || slabBufferCompletionThreshold > 1<<22 {
+		return nil, fmt.Errorf("invalid slabBufferCompletionThreshold %v", slabBufferCompletionThreshold)
+	}
+
+	// load existing buffers
+	var buffers []dbBufferedSlab
+	err := sqlStore.db.
+		Joins("DBSlab").
+		Find(&buffers).
+		Error
+	if err != nil {
+		return nil, err
+	}
+	mgr := &SlabBufferManager{
+		bufferedSlabCompletionThreshold: slabBufferCompletionThreshold,
+		dir:                             partialSlabDir,
+		s:                               sqlStore,
+		completeBuffers:                 make(map[bufferGroupID][]*SlabBuffer),
+		incompleteBuffers:               make(map[bufferGroupID][]*SlabBuffer),
+		buffersByKey:                    make(map[string]*SlabBuffer),
+	}
+	for _, buffer := range buffers {
+		if buffer.DBSlab.ID == 0 {
+			// Buffer doesn't have a slab. We can delete it.
+			if err := sqlStore.db.Delete(&buffer).Error; err != nil {
+				return nil, fmt.Errorf("failed to delete buffer %v: %v", buffer.ID, err)
+			}
+			if err := os.RemoveAll(filepath.Join(partialSlabDir, buffer.Filename)); err != nil {
+				return nil, fmt.Errorf("failed to remove buffer file %v: %v", buffer.Filename, err)
+			}
+			continue
+		}
+		// Open the file.
+		file, err := os.OpenFile(filepath.Join(partialSlabDir, buffer.Filename), os.O_RDWR, 0600)
+		if err != nil {
+			return nil, err
+		}
+		// Create the slab buffer.
+		var ec object.EncryptionKey
+		if err := ec.UnmarshalText(buffer.DBSlab.Key); err != nil {
+			return nil, err
+		}
+		sb := &SlabBuffer{
+			dbID:     buffer.ID,
+			filename: buffer.Filename,
+			slabKey:  ec,
+			maxSize:  int64(bufferedSlabSize(buffer.DBSlab.MinShards)),
+			file:     file,
+			dbSize:   buffer.Size,
+			size:     buffer.Size,
+		}
+		// Add the buffer to the manager.
+		gid := bufferGID(buffer.DBSlab.MinShards, buffer.DBSlab.TotalShards, uint32(buffer.DBSlab.DBContractSetID))
+		if buffer.Complete {
+			mgr.completeBuffers[gid] = append(mgr.completeBuffers[gid], sb)
+		} else {
+			mgr.incompleteBuffers[gid] = append(mgr.completeBuffers[gid], sb)
```

Comment:
this is not right - this should've been flagged by a unit and/or integration test